File: tests/cuda/lua/test_signal_peak_finding.lua
-- Signal peak_finding CUDA binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_peak_finding.lua

local ok_ins, ins = pcall(require, "_insight")
if not ok_ins then
  print("SKIP: _insight module not available")
  os.exit(0)
end

local ok_gpu = pcall(function()
  ins.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

ins.set_device(ins.GPUPlace(0))

describe("Signal Peak Finding CUDA Tests", function()
  it("argrelmax", function()
    local data = ins.from_table({ 0.0, 2.0, 1.0, 3.0, 0.0, 4.0, 1.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmax(data, 1)
    assert.is_not_nil(result)
  end)

  it("argrelmin", function()
    local data = ins.from_table({ 3.0, 1.0, 4.0, 0.0, 5.0, 2.0, 6.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmin(data, 1)
    assert.is_not_nil(result)
  end)

  it("argrelextrema_greater", function()
    local data = ins.from_table({ 0.0, 2.0, 1.0, 3.0, 0.0, 4.0, 1.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelextrema(data, 1)
    assert.is_not_nil(result)
  end)

  it("argrelmax_order2", function()
    local data = ins.from_table({ 0.0, 1.0, 3.0, 2.0, 1.0, 5.0, 2.0, 1.0, 0.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmax(data, 2)
    assert.is_not_nil(result)
  end)

  it("argrelmin_order2", function()
    local data = ins.from_table({ 5.0, 3.0, 1.0, 2.0, 4.0, 0.0, 3.0, 4.0, 5.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmin(data, 2)
    assert.is_not_nil(result)
  end)

  it("argrelmax_no_peaks", function()
    local data = ins.from_table({ 1.0, 2.0, 3.0, 4.0, 5.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmax(data, 1)
    assert.is_not_nil(result)
  end)

  it("argrelmin_no_valleys", function()
    local data = ins.from_table({ 5.0, 4.0, 3.0, 2.0, 1.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmin(data, 1)
    assert.is_not_nil(result)
  end)

  it("argrelmax_sine", function()
    local t = {}
    for i = 0, 99 do
      t[i + 1] = math.sin(4 * math.pi * i / 100)
    end
    local data = ins.from_table(t):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmax(data, 1)
    assert.is_not_nil(result)
  end)

  it("argrelmin_sine", function()
    local t = {}
    for i = 0, 99 do
      t[i + 1] = math.sin(4 * math.pi * i / 100)
    end
    local data = ins.from_table(t):to(ins.GPUPlace(0))
    local result = ins.signal.argrelmin(data, 1)
    assert.is_not_nil(result)
  end)
end)

File: tests/cuda/julia/test_signal_peak_finding.jl
# Signal peak_finding CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_signal_peak_finding.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

# ============================================================================
# Peak Finding
# ============================================================================
println("=== Signal Peak Finding ===")

# argrelmax
data = Insight.from_data([0.0, 2.0, 1.0, 3.0, 0.0, 4.0, 1.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmax(data_gpu, order=1)
check("argrelmax", length(result) >= 0)

# argrelmin
data = Insight.from_data([3.0, 1.0, 4.0, 0.0, 5.0, 2.0, 6.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmin(data_gpu, order=1)
check("argrelmin", length(result) >= 0)

# argrelmax order 2
data = Insight.from_data([0.0, 1.0, 3.0, 2.0, 1.0, 5.0, 2.0, 1.0, 0.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmax(data_gpu, order=2)
check("argrelmax_order2", length(result) >= 0)

# argrelmin order 2
data = Insight.from_data([5.0, 3.0, 1.0, 2.0, 4.0, 0.0, 3.0, 4.0, 5.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmin(data_gpu, order=2)
check("argrelmin_order2", length(result) >= 0)

# argrelmax no peaks (monotonic)
data = Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmax(data_gpu, order=1)
check("argrelmax_no_peaks", length(result) == 0)

# argrelmin no valleys (monotonic)
data = Insight.from_data([5.0, 4.0, 3.0, 2.0, 1.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmin(data_gpu, order=1)
check("argrelmin_no_valleys", length(result) == 0)

# argrelmax sine
sine_data = [sin(4 * pi * i / 100) for i in 0:99]
data = Insight.from_data(sine_data)
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmax(data_gpu, order=1)
check("argrelmax_sine", length(result) > 0)

# argrelmin sine
data = Insight.from_data(sine_data)
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmin(data_gpu, order=1)
check("argrelmin_sine", length(result) > 0)

# argrelmax wider order
sine_data2 = [sin(2 * pi * i / 50) for i in 0:49]
data = Insight.from_data(sine_data2)
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmax(data_gpu, order=3)
check("argrelmax_wide_order", length(result) >= 0)

# argrelmin wider order
data = Insight.from_data(sine_data2)
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
result = Insight.argrelmin(data_gpu, order=3)
check("argrelmin_wide_order", length(result) >= 0)

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
